src/OpenFeature.Contrib.Providers.Flagd/FlagdProvider.cs

```diff
@@ -0,0 +1,375 @@
+        /// </summary>
+        public FlagdProvider(Uri url)
+        {
+            if (url == null)
+            {
+                throw new ArgumentNullException(nameof(url));
+            }
+
+            _client = new Service.ServiceClient(GrpcChannel.ForAddress(url));
+        }
+
+        // just for testing, internal but visible in tests
+        internal FlagdProvider(Service.ServiceClient client)
+        {
+            _client = client;
+        }
+
+        /// <summary>
+        /// Get the provider name.
+        /// </summary>
+        public static string GetProviderName()
+        {
+            return Api.Instance.GetProviderMetadata().Name;
+        }
+
+        /// <summary>
+        ///     Return the metadata associated to this provider.
+        /// </summary>
+        public override Metadata GetMetadata() => _providerMetadata;
+
+        /// <summary>
+        ///     ResolveBooleanValue resolve the value for a Boolean Flag.
+        /// </summary>
+        /// <param name="flagKey">Name of the flag</param>
+        /// <param name="defaultValue">Default value used in case of error.</param>
+        /// <param name="context">Context about the user</param>
+        /// <returns>A ResolutionDetails object containing the value of your flag</returns>
+        public override async Task<ResolutionDetails<bool>> ResolveBooleanValue(string flagKey, bool defaultValue, EvaluationContext context = null)
+        {
+            try
+            {
+                var resolveBooleanResponse = await _client.ResolveBooleanAsync(new ResolveBooleanRequest
+                {
+                    Context = ConvertToContext(context),
+                    FlagKey = flagKey
+                });
+
+                return new ResolutionDetails<bool>(
+                    flagKey: flagKey,
+                    value: resolveBooleanResponse.Value,
+                    reason: resolveBooleanResponse.Reason,
+                    variant: resolveBooleanResponse.Variant
+                );
+            }
+            catch (Grpc.Core.RpcException e)
+            {
+                return GetDefaultWithException<bool>(e, flagKey, defaultValue);
+            }
+        }
+
+        /// <summary>
+        ///     ResolveStringValue resolve the value for a string Flag.
+        /// </summary>
+        /// <param name="flagKey">Name of the flag</param>
+        /// <param name="defaultValue">Default value used in case of error.</param>
+        /// <param name="context">Context about the user</param>
+        /// <returns>A ResolutionDetails object containing the value of your flag</returns>
+        public override async Task<ResolutionDetails<string>> ResolveStringValue(string flagKey, string defaultValue, EvaluationContext context = null)
+        {
+            try
+            {
+                var resolveBooleanResponse = await _client.ResolveStringAsync(new ResolveStringRequest
```

Comment:
that's a good point - I have introduced a generic `ResolveValue` function that accepts a delegate function in which the `_client.GetStringAsync` is being called. It does not reduce a lot of duplications now, apart from the exception handling that only needs to be done within the generic method now, but I believe it will avoid duplications when we add additional logic, e.g. for local caching to the provider. Please let me know if this implementation is somewhat aligned to the idiomatic way of doing things in C#, or if this can be done better :)

Reply:
Ya this is great! We won't have to duplicate caching/logging whatever other side-effects this way.